File: src/lexer.cpp
#include "lexer.h"
#include <cctype>
#include <unordered_map>
#include "token.h"

Lexer::Lexer(const std::string& source) : source(source), position(0) {}

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> tokens;
    while (position < source.length()) {
        char currentChar = source[position];
        if (std::isspace(currentChar)) {
            position++;
        } else if (std::isalpha(currentChar) || currentChar == '_') {
            tokens.push_back(tokenizeIdentifier());
        } else if (std::isdigit(currentChar)) {
            tokens.push_back(tokenizeNumber());
        } else if (currentChar == '"') {
            tokens.push_back(tokenizeString());
        } else {
            switch (currentChar) {
                case '+':
                    tokens.push_back(Token(TokenType::PLUS, "+"));
                    break;
                case '-':
                    tokens.push_back(Token(TokenType::MINUS, "-"));
                    break;
                case '*':
                    tokens.push_back(Token(TokenType::STAR, "*"));
                    break;
                case '/':
                    tokens.push_back(Token(TokenType::SLASH, "/"));
                    break;
                case '=':
                    tokens.push_back(Token(TokenType::EQUAL, "="));
                    break;
                case '(':
                    tokens.push_back(Token(TokenType::LEFT_PAREN, "("));
                    break;
                case ')':
                    tokens.push_back(Token(TokenType::RIGHT_PAREN, ")"));
                    break;
                case '{':
                    tokens.push_back(Token(TokenType::LEFT_BRACE, "{"));
                    break;
                case '}':
                    tokens.push_back(Token(TokenType::RIGHT_BRACE, "}"));
                    break;
                case ':':
                    tokens.push_back(Token(TokenType::COLON, ":"));
                    break;
                case ';':
                    tokens.push_back(Token(TokenType::SEMICOLON, ";"));
                    break;
                case ',':
                    tokens.push_back(Token(TokenType::COMMA, ","));
                    break;
                case '.':
                    tokens.push_back(Token(TokenType::DOT, "."));
                    break;
                default:
                    std::cerr << "Unexpected character: " << currentChar << std::endl;
                    break;
            }
            position++;
        }
    }
    tokens.push_back(Token(TokenType::END_OF_FILE, ""));
    return tokens;
}

Token Lexer::tokenizeIdentifier() {
    size_t start = position;
    while (position < source.length() && (std::isalnum(source[position]) || source[position] == '_')) {
        position++;
    }
    std::string identifier = source.substr(start, position - start);
    TokenType type = TokenType::IDENTIFIER;
    if (keywords.find(identifier) != keywords.end()) {
        type = keywords.at(identifier);
    }
    return Token(type, identifier);
}

Token Lexer::tokenizeNumber() {
    size_t start = position;
    while (position < source.length() && std::isdigit(source[position])) {
        position++;
    }
    std::string number = source.substr(start, position - start);
    return Token(TokenType::NUMBER, number);
}

Token Lexer::tokenizeString() {
    size_t start = position + 1;
    position++;
    while (position < source.length() && source[position] != '"') {
        position++;
    }
    std::string str = source.substr(start, position - start);
    position++;
    return Token(TokenType::STRING, str);
}

const std::unordered_map<std::string, TokenType> Lexer::keywords = {
    {"let", TokenType::LET},
    {"const", TokenType::CONST},
    {"def", TokenType::DEF},
    {"if", TokenType::IF},
    {"else", TokenType::ELSE},
    {"for", TokenType::FOR},
    {"while", TokenType::WHILE},
    {"return", TokenType::RETURN},
    {"class", TokenType::CLASS},
    {"print", TokenType::PRINT}
};

File: src/lexer.ts
enum TokenType {
    BANG_EQUAL,
    IDENTIFIER,
    NUMBER,
    PLUS,
    MINUS,
    STRING,
    STAR,
    SLASH,
    SEMICOLON,
    EQUAL,
    EQUAL_EQUAL,
    LEFT_PAREN,
    RIGHT_PAREN,
    LEFT_BRACE,
    RIGHT_BRACE,
    COLON,
    COMMA,
    DOT,
    LET,
    CONST,
    DEF,
    IF,
    ELSE,
    FOR,
    WHILE,
    RETURN,
    CLASS,
    PRINT,
    END_OF_FILE,
    GREATER,
    GREATER_EQUAL,
    LESS,
    LESS_EQUAL,
    BANG,
    ERROR,
}

class Token {
    constructor(public type: TokenType, public lexeme: string, public literal: string = "") {}
}

class Lexer {
    private position: number = 0;

    private static keywords: { [key: string]: TokenType } = {
        "let": TokenType.LET,
        "const": TokenType.CONST,
        "def": TokenType.DEF,
        "if": TokenType.IF,
        "else": TokenType.ELSE,
        "for": TokenType.FOR,
        "while": TokenType.WHILE,
        "return": TokenType.RETURN,
        "class": TokenType.CLASS,
        "print": TokenType.PRINT
    };

    constructor(private source: string) {}

    public tokenize(): Token[] {
        const tokens: Token[] = [];
        while (this.position < this.source.length) {
            const currentChar = this.source[this.position];
            if (/\s/.test(currentChar)) {
                this.position++;
            } else if (/[a-zA-Z_]/.test(currentChar)) {
                tokens.push(this.tokenizeIdentifier());
            } else if (/\d/.test(currentChar)) {
                tokens.push(this.tokenizeNumber());
            } else if (currentChar === '"') {
                tokens.push(this.tokenizeString());
            } else {
                switch (currentChar) {
                    case '+':
                        tokens.push(new Token(TokenType.PLUS, "+"));
                        break;
                    case '-':
                        tokens.push(new Token(TokenType.MINUS, "-"));
                        break;
                    case '*':
                        tokens.push(new Token(TokenType.STAR, "*"));
                        break;
                    case '/':
                        tokens.push(new Token(TokenType.SLASH, "/"));
                        break;
                    case '=':
                        tokens.push(new Token(TokenType.EQUAL, "="));
                        break;
                    case '(':
                        tokens.push(new Token(TokenType.LEFT_PAREN, "("));
                        break;
                    case ')':
                        tokens.push(new Token(TokenType.RIGHT_PAREN, ")"));
                        break;
                    case '{':
                        tokens.push(new Token(TokenType.LEFT_BRACE, "{"));
                        break;
                    case '}':
                        tokens.push(new Token(TokenType.RIGHT_BRACE, "}"));
                        break;
                    case ':':
                        tokens.push(new Token(TokenType.COLON, ":"));
                        break;
                    case ';':
                        tokens.push(new Token(TokenType.SEMICOLON, ";"));
                        break;
                    case ',':
                        tokens.push(new Token(TokenType.COMMA, ","));
                        break;
                    case '.':
                        tokens.push(new Token(TokenType.DOT, "."));
                        break;
                    default:
                        console.error(`Unexpected character: ${currentChar}`);
                        break;
                }
                this.position++;
            }
        }
        tokens.push(new Token(TokenType.END_OF_FILE, ""));
        return tokens;
    }

    private tokenizeIdentifier(): Token {
        const start = this.position;
        while (this.position < this.source.length && /[a-zA-Z0-9_]/.test(this.source[this.position])) {
            this.position++;
        }
        const identifier = this.source.substring(start, this.position);
        const type = Lexer.keywords[identifier] || TokenType.IDENTIFIER;
        return new Token(type, identifier);
    }

    private tokenizeNumber(): Token {
        const start = this.position;
        while (this.position < this.source.length && /\d/.test(this.source[this.position])) {
            this.position++;
        }
        const number = this.source.substring(start, this.position);
        return new Token(TokenType.NUMBER, number);
    }

    private tokenizeString(): Token {
        const start = this.position + 1;
        this.position++;
        while (this.position < this.source.length && this.source[this.position] !== '"') {
            this.position++;
        }
        const str = this.source.substring(start, this.position);
        this.position++;
        return new Token(TokenType.STRING, str);
    }
}

export { TokenType, Token, Lexer };
